feat(plugin-chart-echarts): subject Add option to rotate label defaul 45

maloun96 opened the pull request:
Add option to rotate x axis label

Close https://github.com/apache/superset/issues/12775

## After
https://user-images.githubusercontent.com/8277264/106272176-4e39c980-6239-11eb-947e-839b129e4c43.mp4

villebro commented on plugins/plugin-chart-echarts/src/Timeseries/controlPanel.tsx:
```diff
@@ -338,6 +339,18 @@ const config: ControlPanelConfig = {
+        [
+          {
+            name: 'xAxisLabelRotation',
+            config: {
+              type: 'TextControl',
+              label: t('Rotate x axis label'),
+              default: xAxisLabelRotation,
+              renderTrigger: true,
+              description: t('Rotate x axis label value between -90 and 90'),
```
Let's add a number validator here. Also we should mention that it's degrees, something like
```
              description: t('Rotation of x axis label in degrees'),
```

villebro commented on plugins/plugin-chart-echarts/src/Timeseries/transformProps.ts:
```diff
@@ -153,6 +154,10 @@ export default function transformProps(chartProps: ChartProps): EchartsProps {
         showMinLabel: xAxisShowMinLabel,
         showMaxLabel: xAxisShowMaxLabel,
         formatter: xAxisFormatter,
+        rotate:
+          xAxisLabelRotation >= -90 && xAxisLabelRotation <= 90
+            ? xAxisLabelRotation
+            : DEFAULT_FORM_DATA.xAxisLabelRotation,
```
I think we can remove this - let's do as much validation as possible in the control. I think validating for numeric values is enough at this point (in reference to the previous comment).

ktmud replied:
We should probably allow other degrees if ECharts supports it. I'm always in favor of not adding hurdles to strange user behaviors unless absolutely necessary.

ktmud commented on plugins/plugin-chart-echarts/src/Timeseries/controlPanel.tsx:
```diff
@@ -338,6 +339,18 @@ const config: ControlPanelConfig = {
+          {
+            name: 'xAxisLabelRotation',
+            config: {
+              type: 'TextControl',
```
Maybe use `SelectControl` to give users the most common options? (Should have `freeForm` enabled so users can use other values).

maloun96 commented:
https://user-images.githubusercontent.com/8277264/106379889-af87a700-63b7-11eb-8f01-140d49cb305a.mp4

So I removed the interval and added an integer validation, I think it's ok, user can introduce any number and chart il rotate label (see video)

ktmud commented:
Still think free form `SelectControl` is a better choice, which is in line with other controls such as row limit, page length, and these options in Bar Chart: 

<img width="404" alt="other-select" src="https://user-images.githubusercontent.com/335541/106430416-9bc45b00-6420-11eb-8ed0-b044f9829bd3.png">

junlincc commented:
when we have long label names, 45 degree is gonna take a lot of space.  @ktmud

junlincc commented:
auto, 45°, free form. and tooltip " Input field supports custom rotation. e.g. 30 for 30° "

maloun96 commented:
UPDATE:
Added select control with free form, default is 45 degrees
https://user-images.githubusercontent.com/8277264/106441208-710aee80-6482-11eb-8ddf-154d21948493.mp4

ktmud commented on plugins/plugin-chart-echarts/src/Timeseries/controlPanel.tsx:
```diff
@@ -338,6 +339,24 @@ const config: ControlPanelConfig = {
+              label: t('Rotate x axis label'),
+              choices: [
+                [0, '0°'],
+                [45, '45°'],
```
Maybe add `90`, too?

maloun96 replied:
No need for 90 after discussion with jun

ktmud replied:
That's fine, I'm not a fan of 90 degree rotation, either.

But I just wanna point out that it is actually used in the official examples of [HighCharts](https://api.highcharts.com/highcharts/xAxis.labels.rotation) and [AMCharts](https://www.amcharts.com/demos/column-with-rotated-series/), so I'd imagine some users may actually prefer it for some cases.

junlincc replied:
![tenor](https://user-images.githubusercontent.com/67837651/106987501-75553700-6722-11eb-942a-b17853b46548.gif)

ktmud replied:
No they won't. ECharts API does not allow that. [ECharts options only allow -90 to 90.](https://echarts.apache.org/en/option.html#xAxis.axisLabel.rotate)

ktmud replied:
We should probably add a custom validator for this, btw. Not super important, though.
